Blocks/Persistence/file.h

```diff
@@ -426,117 +425,128 @@ class FilePersister {
+    file_persister_impl_->record_offset_.push_back(file_persister_impl_->file_appender_.tellp());
+    file_persister_impl_->record_timestamp_.push_back(timestamp);
+
+    // Explicit `ConstructAsNecessary` is essential, otherwise the `Variant`'s case
+    // would be serialized in an unwrapped way when passed directly
+    file_persister_impl_->file_appender_ << JSON(idxts) << '\t'
+                                         << JSON(ConstructAsNecessary<ENTRY, decay<E>>::DoIt(std::forward<E>(entry)))
+                                         << std::endl;
     ++iterator.next_index;
-    file_persister_impl_->head_offset = 0;
-    file_persister_impl_->end.store(iterator);
+    file_persister_impl_->head_offset_ = 0;
+    file_persister_impl_->end_.store(iterator);
 
-    return current;
+    return idxts;
   }
 
-  template <current::locks::MutexLockStatus MLS, typename US>
-  void DoUpdateHead(const US us) {
-    current::locks::SmartMutexLockGuard<MLS> lock(file_persister_impl_->mutex_ref);
+  template <current::locks::MutexLockStatus MLS, typename TIMESTAMP>
+  void PersisterUpdateHeadImpl(TIMESTAMP&& provided_timestamp) {
+    current::locks::SmartMutexLockGuard<MLS> lock(file_persister_impl_->publish_mutex_ref_);
 
-    end_t iterator = file_persister_impl_->end.load();
-    const auto timestamp = current::time::GetTimestampFromLockedSection(us);
+    end_t iterator = file_persister_impl_->end_.load();
+    const auto timestamp = current::time::GetTimestampFromLockedSection(std::forward<TIMESTAMP>(provided_timestamp));
     if (!(timestamp > iterator.head)) {
       CURRENT_THROW(ss::InconsistentTimestampException(iterator.head + std::chrono::microseconds(1), timestamp));
     }
     iterator.head = timestamp;
     const auto head_str = Printf(constants::kHeadFormatString, static_cast<long long>(timestamp.count()));
-    if (file_persister_impl_->head_offset) {
-      auto& rewriter = file_persister_impl_->head_rewriter;
-      rewriter.seekp(file_persister_impl_->head_offset, std::ios_base::beg);
+    if (file_persister_impl_->head_offset_) {
+      auto& rewriter = file_persister_impl_->head_rewriter_;
+      rewriter.seekp(file_persister_impl_->head_offset_, std::ios_base::beg);
       rewriter << head_str << std::endl;
     } else {
-      auto& appender = file_persister_impl_->appender;
-      appender << constants::kHeadDirective << ' ';
-      file_persister_impl_->head_offset = appender.tellp();
-      appender << head_str << std::endl;
+      auto& file_appender_ = file_persister_impl_->file_appender_;
+      file_appender_ << constants::kHeadDirective << ' ';
+      file_persister_impl_->head_offset_ = file_appender_.tellp();
+      file_appender_ << head_str << std::endl;
     }
-    file_persister_impl_->end.store(iterator);
+    file_persister_impl_->end_.store(iterator);
+  }
+
+  template <current::locks::MutexLockStatus MLS>
+  bool PersisterEmptyImpl() const noexcept {
```

Comment:
Strictly speaking, using `noexcept` here and below is wrong, as our atomic's `load()` constructs a `lock_guard` and may throw.